packages/dds/tree/docs/wip/cell-model.md

```diff
@@ -0,0 +1,210 @@
+
+It's simpler to start with a foundation of tree locations (cells) that never move/disappear, and then layer the idea of moveable/removable nodes on top of that.
+
+The impedance mismatch between our internal model and the semantics desired by users was making things unnecessarily complicated: forcing the internal model to be excessively general, and then requiring higher layers to compensate.
+The primary example being the use of constraints and/or hierarchical edits to constrain the semantics of insert/remove to fit the desired semantics of setting a fixed-sized field.
+
+## Motivation
+
+We present here an example that portrays the challenges encountered by editing models based on insertion and removal.
+
+Consider a collaborative editing session between three participants (all acting concurrently), over a field populated with elements [A, B, C, D], where the edits are sequenced as follows:
+
+1. User 1: remove B, C (local state: [A, D])
+
+2. User 2: insert Y after C (local state: [A, B, C, Y, D])
+
+3. User 3: insert X after B (local state: [A, B, X, C, D])
+
+4. User 1: undo the removal of B, C (local state: [A, B, C, D])
+
+The outcome we expect from such a session is [A, B, X, C, Y, D].
+
+Traditional models based on insertion and removal may fail to produce this outcome in two ways:
+
+* The order of X relative to Y may not be correct (i.e., Y may appear before X in the list of elements).
+
+* The order the inserted element relative to the temporarily removed elements may not be correct (yielding either [A, X, Y, B, C, D] or [A, B, C, X, Y, D]).
+
+The first failure stems from the fact that, as the insertion of X and Y are made relative to removed content, information is lost as to their precise insertion point.
+
+The second failure stems from the fact that the undo of the removal of B, C by user 1 is modeled as an insertion.
+
+This second failure can also occur when elements are moved-out and the move is then undone (and the undo is modeled as a move in the opposite direction).
+
+These failures can be addressed by introducing new features:
+
+* Tombstones, which represent removed or moved-out elements
+
+* The "revive" operation which restores elements in the place of their tombstones
+
+* The "return" operation which moves elements back in the place of their tombstone.
+
+These additions, while effective, have the following drawbacks:
+
+* They increase the complexity of the core editing model
+- They lead to a set of operations that is not orthogonalized:
+  
+  - Insert, revive, and return seem to share some common traits, with revive and return being closer to each other.
+  
+  - It seems like it should be possible to support a "replace" operation that is similar to revive but with new elements instead of resurrecting removed ones, but that too needs to be added as separate operation)
+
+- They do not alleviate the complexity of modeling an optional field.
```

Comment:
Still relatively new to this problem space, but it isn't immediately clear to me what the complexities here would be. An example or two might help 🙂

Reply:
I added a comment to clarify.